File: lua/mpp/protocol/core/headers.lua
local challenge = require('mpp.protocol.core.challenge')
local types = require('mpp.protocol.core.types')
local json = require('mpp.util.json')

local M = {
  WWW_AUTHENTICATE_HEADER = 'www-authenticate',
  AUTHORIZATION_HEADER = 'authorization',
  PAYMENT_RECEIPT_HEADER = 'payment-receipt',
  PAYMENT_SCHEME = 'Payment',
}

local max_token_len = 16 * 1024

local function escape_quoted(value)
  value = tostring(value)
  value = value:gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\r', ''):gsub('\n', '')
  return value
end

local function strip_payment_scheme(header)
  local trimmed = header:gsub('^%s+', '')
  if trimmed:sub(1, #M.PAYMENT_SCHEME):lower() ~= string.lower(M.PAYMENT_SCHEME) then
    return nil
  end
  return trimmed:sub(#M.PAYMENT_SCHEME + 1):gsub('^%s+', '')
end

local function parse_auth_params(input)
  local params = {}
  local i = 1
  while i <= #input do
    while i <= #input and input:sub(i, i):match('[%s,]') do
      i = i + 1
    end
    if i > #input then
      break
    end
    local eq = input:find('=', i, true)
    if not eq then
      error('invalid auth parameter')
    end
    local key = input:sub(i, eq - 1):gsub('%s+$', '')
    i = eq + 1
    local value
    if input:sub(i, i) == '"' then
      i = i + 1
      local out = {}
      local escaped = false
      while i <= #input do
        local ch = input:sub(i, i)
        i = i + 1
        if escaped then
          out[#out + 1] = ch
          escaped = false
        elseif ch == '\\' then
          escaped = true
        elseif ch == '"' then
          break
        else
          out[#out + 1] = ch
        end
      end
      value = table.concat(out)
    else
      local next_comma = input:find(',', i, true)
      if next_comma then
        value = input:sub(i, next_comma - 1):gsub('%s+$', '')
        i = next_comma + 1
      else
        value = input:sub(i):gsub('%s+$', '')
        i = #input + 1
      end
    end
    if params[key] ~= nil then
      error('duplicate parameter: ' .. key)
    end
    params[key] = value
  end
  return params
end

function M.extract_payment_scheme(header)
  for part in header:gmatch('[^,]+') do
    local trimmed = part:gsub('^%s+', ''):gsub('%s+$', '')
    if trimmed:sub(1, #M.PAYMENT_SCHEME + 1):lower() == string.lower(M.PAYMENT_SCHEME) .. ' ' then
      return trimmed
    end
  end
  return nil
end

function M.parse_www_authenticate(header)
  local rest = strip_payment_scheme(header)
  if not rest then
    error('expected "Payment" scheme')
  end
  local params = parse_auth_params(rest)
  if not params.request or params.request == '' then
    error('missing "request" field')
  end
  local request_bytes, decode_err = types.base64url_decode(params.request)
  if not request_bytes then
    error('invalid request field: ' .. decode_err)
  end
  local ok = pcall(json.decode, request_bytes)
  if not ok then
    error('invalid JSON in request field')
  end
  local method = types.new_method_name(params.method or '')
  if not types.is_valid_method(method) then
    error('invalid method: ' .. tostring(params.method))
  end
  if not params.id or params.id == '' or not params.realm or params.realm == '' or not params.intent or params.intent == '' then
    error('missing required challenge fields')
  end
  return challenge.challenge_from_table({
    id = params.id,
    realm = params.realm,
    method = method,
    intent = types.new_intent_name(params.intent),
    request = params.request,
    expires = params.expires,
    description = params.description,
    digest = params.digest,
    opaque = params.opaque,
  })
end

function M.format_www_authenticate(value)
  local plain = challenge.challenge_to_plain(value)
  local parts = {
    'id="' .. escape_quoted(plain.id) .. '"',
    'realm="' .. escape_quoted(plain.realm) .. '"',
    'method="' .. escape_quoted(plain.method) .. '"',
    'intent="' .. escape_quoted(plain.intent) .. '"',
    'request="' .. escape_quoted(plain.request) .. '"',
  }
  if plain.expires and plain.expires ~= '' then
    parts[#parts + 1] = 'expires="' .. escape_quoted(plain.expires) .. '"'
  end
  if plain.description and plain.description ~= '' then
    parts[#parts + 1] = 'description="' .. escape_quoted(plain.description) .. '"'
  end
  if plain.digest and plain.digest ~= '' then
    parts[#parts + 1] = 'digest="' .. escape_quoted(plain.digest) .. '"'
  end
  if plain.opaque and plain.opaque ~= '' then
    parts[#parts + 1] = 'opaque="' .. escape_quoted(plain.opaque) .. '"'
  end
  return M.PAYMENT_SCHEME .. ' ' .. table.concat(parts, ', ')
end

function M.parse_authorization(header)
  local token = M.extract_payment_scheme(header)
  if not token then
    error('expected "Payment" scheme')
  end
  token = token:sub(#M.PAYMENT_SCHEME + 1):gsub('^%s+', '')
  if #token > max_token_len then
    error('token exceeds maximum length of ' .. max_token_len .. ' bytes')
  end
  local payload, decode_err = types.base64url_decode(token)
  if not payload then
    error(decode_err)
  end
  local ok, value = pcall(json.decode, payload)
  if not ok then
    error('invalid credential JSON: ' .. value)
  end
  value.challenge = challenge.challenge_from_table(value.challenge)
  return value
end

function M.format_authorization(value)
  local payload = json.encode(challenge.credential_to_plain(value))
  return M.PAYMENT_SCHEME .. ' ' .. types.base64url_encode(payload)
end

function M.parse_receipt(header)
  if #header > max_token_len then
    error('receipt exceeds maximum length of ' .. max_token_len .. ' bytes')
  end
  local payload, decode_err = types.base64url_decode((header:gsub('^%s+', ''):gsub('%s+$', '')))
  if not payload then
    error(decode_err)
  end
  local ok, value = pcall(json.decode, payload)
  if not ok then
    error('invalid receipt JSON: ' .. value)
  end
  return value
end

function M.format_receipt(value)
  return types.base64url_encode(json.encode(challenge.receipt_to_plain(value)))
end

return M

File: python/src/solana_mpp/_headers.py
"""Header parsing and formatting for Web Payment Auth.

No regex for auth-param parsing -- follows the same hand-rolled parser as the
Rust and Go implementations.
"""

from __future__ import annotations

import json
from typing import Any

from solana_mpp._base64url import decode, decode_json, encode_json
from solana_mpp._types import ChallengeEcho, PaymentChallenge, PaymentCredential, Receipt

MAX_TOKEN_LEN = 16 * 1024

PAYMENT_SCHEME = "Payment"
WWW_AUTHENTICATE_HEADER = "www-authenticate"
AUTHORIZATION_HEADER = "authorization"
PAYMENT_RECEIPT_HEADER = "payment-receipt"


class ParseError(Exception):
    """Failed to parse a payment header."""


# ---------------------------------------------------------------------------
# WWW-Authenticate
# ---------------------------------------------------------------------------


def parse_www_authenticate(header: str) -> PaymentChallenge:
    """Parse a WWW-Authenticate header into a PaymentChallenge."""
    rest = _strip_payment_scheme(header)
    if rest is None:
        raise ParseError("Expected 'Payment' scheme")

    # Must have a space or tab separator
    if not rest or (rest[0] != " " and rest[0] != "\t"):
        raise ParseError("Expected space after 'Payment' scheme")

    params = _parse_auth_params(rest.lstrip())

    id_ = _require_param(params, "id")
    if not id_:
        raise ParseError("Empty 'id' parameter")

    realm = _require_param(params, "realm")
    method_raw = _require_param(params, "method")
    if not method_raw or not all(c.islower() and c.isascii() for c in method_raw):
        raise ParseError(f'Invalid method: "{method_raw}". Must be lowercase ASCII.')

    intent = _require_param(params, "intent")
    request_b64 = _require_param(params, "request")

    # Validate that request is valid base64url JSON
    try:
        request_bytes = decode(request_b64)
        json.loads(request_bytes)
    except Exception as exc:
        raise ParseError(f"Invalid JSON in request field: {exc}") from exc

    return PaymentChallenge(
        id=id_,
        realm=realm,
        method=method_raw,
        intent=intent,
        request=request_b64,
        expires=params.get("expires", ""),
        description=params.get("description", ""),
        digest=params.get("digest", ""),
        opaque=params.get("opaque"),
    )


def format_www_authenticate(challenge: PaymentChallenge) -> str:
    """Format a PaymentChallenge as a WWW-Authenticate header value."""
    parts = [
        f'id="{_escape_quoted_value(challenge.id)}"',
        f'realm="{_escape_quoted_value(challenge.realm)}"',
        f'method="{_escape_quoted_value(challenge.method)}"',
        f'intent="{_escape_quoted_value(challenge.intent)}"',
        f'request="{_escape_quoted_value(challenge.request)}"',
    ]
    if challenge.expires:
        parts.append(f'expires="{_escape_quoted_value(challenge.expires)}"')
    # description is already encoded inside the `request` payload —
    # don't duplicate it as a top-level header param (non-ASCII descriptions
    # would make the header value invalid).
    if challenge.digest:
        parts.append(f'digest="{_escape_quoted_value(challenge.digest)}"')
    if challenge.opaque is not None:
        parts.append(f'opaque="{_escape_quoted_value(challenge.opaque)}"')

    return f"Payment {', '.join(parts)}"


# ---------------------------------------------------------------------------
# Authorization
# ---------------------------------------------------------------------------


def parse_authorization(header: str) -> PaymentCredential:
    """Parse an Authorization header into a PaymentCredential."""
    payment_part = _extract_payment_scheme(header)
    if payment_part is None:
        raise ParseError("Expected 'Payment' scheme")

    token = payment_part[8:].strip() if len(payment_part) > 8 else ""

    if len(token) > MAX_TOKEN_LEN:
        raise ParseError(f"Token exceeds maximum length of {MAX_TOKEN_LEN} bytes")

    try:
        decoded = decode(token)
        data = json.loads(decoded)
    except Exception as exc:
        raise ParseError(f"Invalid credential: {exc}") from exc

    if not isinstance(data, dict) or "challenge" not in data:
        raise ParseError("Invalid credential JSON structure")

    ch = data["challenge"]
    echo = ChallengeEcho(
        id=str(ch.get("id", "")),
        realm=str(ch.get("realm", "")),
        method=str(ch.get("method", "")),
        intent=str(ch.get("intent", "")),
        request=str(ch.get("request", "")),
        expires=str(ch.get("expires", "")),
        digest=str(ch.get("digest", "")),
        opaque=ch.get("opaque"),
    )

    return PaymentCredential(
        challenge=echo,
        payload=data.get("payload", {}),
        source=data.get("source"),
    )


def format_authorization(credential: PaymentCredential) -> str:
    """Format a PaymentCredential as an Authorization header value."""
    challenge_dict: dict[str, Any] = {
        "challenge": {
            "id": credential.challenge.id,
            "intent": credential.challenge.intent,
            "method": credential.challenge.method,
            "realm": credential.challenge.realm,
            "request": credential.challenge.request,
        },
        "payload": credential.payload,
    }
    # Add optional challenge fields
    ch_dict = challenge_dict["challenge"]
    if credential.challenge.expires:
        ch_dict["expires"] = credential.challenge.expires
    if credential.challenge.digest:
        ch_dict["digest"] = credential.challenge.digest
    if credential.challenge.opaque is not None:
        ch_dict["opaque"] = credential.challenge.opaque

    if credential.source:
        challenge_dict["source"] = credential.source

    encoded = encode_json(challenge_dict)
    return f"Payment {encoded}"


# ---------------------------------------------------------------------------
# Payment-Receipt
# ---------------------------------------------------------------------------


def parse_receipt(header: str) -> Receipt:
    """Parse a Payment-Receipt header into a Receipt."""
    token = header.strip()
    if len(token) > MAX_TOKEN_LEN:
        raise ParseError(f"Receipt exceeds maximum length of {MAX_TOKEN_LEN} bytes")

    try:
        data = decode_json(token)
    except Exception as exc:
        raise ParseError(f"Invalid receipt: {exc}") from exc

    if not isinstance(data, dict):
        raise ParseError("Invalid receipt JSON structure")

    return Receipt(
        status=str(data.get("status", "")),
        method=str(data.get("method", "")),
        timestamp=str(data.get("timestamp", "")),
        reference=str(data.get("reference", "")),
        challenge_id=str(data.get("challengeId", "")),
        external_id=str(data.get("externalId", "")),
    )


def format_receipt(receipt: Receipt) -> str:
    """Format a Receipt as a Payment-Receipt header value."""
    data: dict[str, Any] = {
        "challengeId": receipt.challenge_id,
        "method": receipt.method,
        "reference": receipt.reference,
        "status": receipt.status,
        "timestamp": receipt.timestamp,
    }
    if receipt.external_id:
        data["externalId"] = receipt.external_id
    return encode_json(data)


# ---------------------------------------------------------------------------
# Private helpers
# ---------------------------------------------------------------------------


def _strip_payment_scheme(header: str) -> str | None:
    """Strip the 'Payment' scheme prefix, returning the rest or None."""
    header = header.lstrip()
    if len(header) < len(PAYMENT_SCHEME):
        return None
    if header[: len(PAYMENT_SCHEME)].lower() != PAYMENT_SCHEME.lower():
        return None
    return header[len(PAYMENT_SCHEME) :]


def _extract_payment_scheme(header: str) -> str | None:
    """Extract the Payment scheme section from a comma-separated header."""
    for part in header.split(","):
        part = part.strip()
        if len(part) >= 8 and part[:8].lower() == "payment ":
            return part
    return None


def _escape_quoted_value(s: str) -> str:
    """Escape a string for use in a quoted-string header value."""
    if "\r" in s or "\n" in s:
        raise ParseError("Header value contains invalid CRLF characters")
    return s.replace("\\", "\\\\").replace('"', '\\"')


def _require_param(params: dict[str, str], key: str) -> str:
    """Require a parameter is present in the params dict."""
    if key not in params:
        raise ParseError(f"Missing '{key}' field")
    return params[key]


def _parse_auth_params(params_str: str) -> dict[str, str]:
    """Parse key=value or key="value" pairs from an auth-params string."""
    params: dict[str, str] = {}
    chars = list(params_str)
    i = 0

    while i < len(chars):
        # Skip whitespace and commas
        while i < len(chars) and (chars[i] in (" ", "\t", ",")):
            i += 1
        if i >= len(chars):
            break

        # Read key
        key_start = i
        while i < len(chars) and chars[i] != "=" and chars[i] not in (" ", "\t"):
            i += 1
        if i >= len(chars) or chars[i] != "=":
            # Skip to next comma or whitespace
            while i < len(chars) and chars[i] not in (" ", "\t", ","):
                i += 1
            continue

        key = "".join(chars[key_start:i])
        i += 1  # skip '='

        if i >= len(chars):
            break

        # Read value
        if chars[i] == '"':
            i += 1  # skip opening quote
            value_parts: list[str] = []
            while i < len(chars) and chars[i] != '"':
                if chars[i] == "\\" and i + 1 < len(chars):
                    i += 1
                    value_parts.append(chars[i])
                else:
                    value_parts.append(chars[i])
                i += 1
            if i < len(chars):
                i += 1  # skip closing quote
            value = "".join(value_parts)
        else:
            value_start = i
            while i < len(chars) and chars[i] not in (" ", "\t", ","):
                i += 1
            value = "".join(chars[value_start:i])

        if key in params:
            raise ParseError(f"Duplicate parameter: {key}")
        params[key] = value

    return params
